### PR TITLE
misc(shwap/store): miscelaneous improvements 

### PR DESCRIPTION
**Pull request by walldiss**
Contains five commits with various bug fixes and improvements.

The PR is meant to be merged as 5 separate commits grouped under a common theme.

### CODE REVIEW

File: store/file/ods_q4.go [hunk 1]
```diff
@@ -69,32 +70,35 @@ func OpenODSQ4(pathODS, pathQ4 string) (*ODSQ4, error) {
...
+
+	q4, err := OpenQ4(odsq4.pathQ4)
+	if errors.Is(err, os.ErrNotExist) {
+		return
 	}
 
-	q4, err := odsq4.q4Open()
 	if err != nil {
-		return nil, fmt.Errorf("failed to open Q4: %w", err)
+		log.Errorf("opening Q4 file %s: %s", odsq4.pathQ4, err)
```
Comment (Wondertan) [inline store/file/ods_q4.go:89]: shouldn't we handle those?

File: store/store.go [hunk 2]
```diff
@@ -228,34 +223,42 @@ func (s *Store) GetByHeight(ctx context.Context, height uint64) (eds.AccessorStr
...
 
-// openODSQ4 opens ODSQ4 Accessor.
+// openFile opens ODSQ4 or ODS Accessor.
 // It opens ODS file first, reads up its DataHash and constructs the path for Q4
 // This done as Q4 is not indexed(hard-linked) and there is no other way to Q4 by height only.
-func (s *Store) openODSQ4(path string) (eds.AccessorStreamer, error) {
-	odsq4, err := file.OpenODSQ4(path+odsFileExt, path+q4FileExt)
-	switch {
-	case errors.Is(err, os.ErrNotExist):
+func (s *Store) openODSQ4(ctx context.Context, path string) (eds.AccessorStreamer, error) {
```
Comment (Wondertan) [inline store/store.go:244]: name and comment are inconsistent.

`openAcccesor`?

File: store/file/q4.go [hunk 3]
```diff
@@ -102,17 +89,12 @@ func writeQ4(w io.Writer, eds *rsmt2d.ExtendedDataSquare) error {
 // of the File is read lazily.
 // If file is empty, the ErrEmptyFile is returned.
 // File must be closed after usage.
-func OpenQ4(path string) (*Q4, error) {
+func OpenQ4(path string, hdr *headerV0) (*Q4, error) {
```
Comment (Wondertan) [inline store/file/q4.go:92]: It feels wrong to pass private header. I propose to either:
* Make Header public, doesn't hurt.
* Make OpenQ4 private
  * Then stop implementing Accessor for it. It's not needed anymore as its not usable on its own.

File: store/file/ods_q4.go (hunk 1 above)
Comment (walldiss) [reply store/file/ods_q4.go:89]: If we fail to open Q4, we still allow reader to read from Q1

File: store/file/header.go [hunk 4]
```diff
@@ -8,16 +8,14 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+const headerVOSize int = 40
```
[the remark this answers is not in the document]
Comment (Wondertan) [reply store/file/header.go:11]: Sync: agreed to preallocate 64 bytes for future header additional fields

File: store/file/q4.go (hunk 3 above)
Comment (walldiss) [reply store/file/q4.go]: implemented option 2

File: store/file/header.go [hunk 5]
```diff
@@ -8,16 +8,14 @@ import (
 	"github.com/celestiaorg/celestia-node/share"
 )
 
+const headerVOSize int = 64
```
Comment (Wondertan) [inline store/file/header.go:11]: There must be a note explaining why there is more data than needed

File: store/file/ods_q4.go [hunk 6]
```diff
@@ -60,41 +61,36 @@ func CreateODSQ4(
...
 	if odsq4.q4Opened.Load() {
-		return odsq4.q4, nil
+		return
 	}
 
 	odsq4.q4Mu.Lock()
 	defer odsq4.q4Mu.Unlock()
-	if odsq4.q4Opened.Load() {
-		return odsq4.q4, nil
+	if loaded := odsq4.q4Opened.Swap(true); loaded {
```
Comment (Wondertan) [inline store/file/ods_q4.go:79]: Imo, this needs a note anyway on why its Swap and then we can avoid creating loaded

File: store/file/q4.go [hunk 7]
```diff
@@ -141,30 +119,6 @@ func (q4 *Q4) AxisHalf(ctx context.Context, axisType rsmt2d.Axis, axisIdx int) (
 	}, nil
 }
 
-func (q4 *Q4) Size(context.Context) int {
+func (q4 *q4) Size(context.Context) int {
```
Comment (Wondertan) [inline store/file/q4.go:122]: Just let it access directly from header method

File: store/file/q4.go [hunk 8]
```diff
@@ -97,33 +80,28 @@ func writeQ4(w io.Writer, eds *rsmt2d.ExtendedDataSquare) error {
 	return nil
 }
 
-// OpenQ4 opens an existing Q4 file under given FS path.
+// openQ4 opens an existing Q4 file under given FS path.
```
Comment (Wondertan) [inline store/file/q4.go:83]: Needs an update that it doesnt read the header

File: store/file/q4.go [hunk 9]
```diff
@@ -10,29 +10,24 @@ import (
...
 
-	"github.com/celestiaorg/celestia-node/share"
 	eds "github.com/celestiaorg/celestia-node/share/new_eds"
-	"github.com/celestiaorg/celestia-node/share/shwap"
 )
 
-var _ eds.AccessorStreamer = (*Q4)(nil)
-
-// Q4 implements eds.Accessor as an FS file.
+// q4 implements eds.Accessor as an FS file.
```
Comment (Wondertan) [inline store/file/q4.go:16]: It doesnt implement anymore

File: store/file/ods_q4.go [hunk 10]
```diff
@@ -24,10 +25,10 @@ var _ eds.AccessorStreamer = (*ODSQ4)(nil)
 type ODSQ4 struct {
 	ods *ODS
 
-	q4Open   func() (*Q4, error)
 	q4Mu     sync.Mutex
 	q4Opened atomic.Bool
-	q4       *Q4
+	pathQ4   string
```
Comment (Wondertan) [inline store/file/ods_q4.go:30]: Uber nit: put above mutex, as mutex doesnt protect it.

File: store/file/ods_q4.go [hunk 11]
```diff
@@ -125,22 +121,15 @@ func (odsq4 *ODSQ4) Sample(ctx context.Context, rowIdx, colIdx int) (shwap.Sampl
 func (odsq4 *ODSQ4) AxisHalf(ctx context.Context, axisType rsmt2d.Axis, axisIdx int) (eds.AxisHalf, error) {
 	size := odsq4.Size(ctx) // TODO(@Wondertan): Should return error.
 
-	var acsr eds.Accessor = odsq4.ods
 	if axisIdx >= size/2 {
-		q4, err := odsq4.getQ4()
-		if err != nil {
-			return eds.AxisHalf{}, err
+		// lazy load Q4 file and read axis from it if loaded
+		odsq4.tryLoadQ4()
```
Comment (Wondertan) [inline store/file/ods_q4.go:126]: Readability nit: i would still make it return q4, so it doesn't confuse readers with accessing Q4 field without synchronization, while it has it on the struct. Returning it makes it seem like the tryLoadq4 takes full responsibility on proper synchornization

File: store/file/q4.go (hunk 7 above)
Comment (Wondertan) [reply store/file/q4.go:122]: Actually, that size check isn't really needed there, as it happens on ODSQ4 and Q4 isnt general anymore

File: store/file/ods_q4.go [hunk 12]
```diff
@@ -60,41 +63,39 @@ func CreateODSQ4(
...
 	if odsq4.q4Opened.Load() {
-		return odsq4.q4, nil
+		return odsq4.q4
 	}
 
 	odsq4.q4Mu.Lock()
 	defer odsq4.q4Mu.Unlock()
-	if odsq4.q4Opened.Load() {
-		return odsq4.q4, nil
+	// check again in case another goroutine opened the file before we acquired the lock.
```
Comment (Wondertan) [inline store/file/ods_q4.go:82]:
```suggestion
	// update bool to make sure we try opening only once
	// no matter if the try was successful or failed.
```

File: store/file/ods_q4.go [hunk 13]
```diff
@@ -61,21 +63,15 @@ func CreateODSQ4(
...
 
-// OpenODSQ4 opens ODS file under the given FS path. The Q4 is opened lazily
-// on demand.
-func OpenODSQ4(pathODS, pathQ4 string) (*ODSQ4, error) {
-	ods, err := OpenODS(pathODS)
-	if err != nil {
-		return nil, fmt.Errorf("failed to open ODS: %w", err)
-	}
-
+// ODSWithQ4 returns ODSQ4 instance for the given FS paths. It opens Q4 file lazily.
```
Comment (Wondertan) [inline store/file/ods_q4.go:66]:
```suggestion
// ODSWithQ4 returns ODSQ4 instance over ODS. It opens Q4 file lazily under the given path.
```

**Review by Wondertan — COMMENTED** [review]
Can also remove `Offset` method from header in the last commit

**Pull request merged**